File: Cpp/BaekJoon/Gold/G2_20183.cpp
#include <climits>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Node {
  int node;
  long long dist;

  bool operator<(const Node& right) const {
    if (dist != right.dist) return dist > right.dist;

    return false;
  }
};

bool bfs(int n, const vector<vector<Node>>& graph, long long mid, int a, int b,
         long long c) {
  vector<long long> dist(n + 1, LLONG_MAX);

  priority_queue<Node> pque;
  pque.push({a, 0});

  dist[a] = 0;

  while (!pque.empty()) {
    Node now = pque.top();
    pque.pop();

    if (now.dist > dist[now.node]) continue;

    for (size_t i = 0; i < graph[now.node].size(); i++) {
      Node next = graph[now.node][i];

      if (next.dist > mid) continue;

      long long next_dist = next.dist + now.dist;

      if (next_dist > c) continue;
      if (next_dist >= dist[next.node]) continue;

      pque.push({next.node, next_dist});
      dist[next.node] = next_dist;
    }
  }

  if (dist[b] != LLONG_MAX) {
    return true;
  } else {
    return false;
  }
}

int main() {
  int n, m, a, b;
  long long c;

  cin >> n >> m >> a >> b >> c;

  vector<vector<Node>> graph(n + 1);

  int max_cost = 0;
  for (int i = 0; i < m; i++) {
    int u, v;
    long long cost;

    cin >> u >> v >> cost;

    if (cost > max_cost) max_cost = cost;

    graph[u].push_back({v, cost});
    graph[v].push_back({u, cost});
  }

  long long lo = 1;
  long long hi = max_cost;

  long long result = -1;

  while (lo <= hi) {
    long long mid = (lo + hi) / 2;
    bool res = bfs(n, graph, mid, a, b, c);

    if (res) {
      result = mid;

      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }

  cout << result;

  return 0;
}

File: Java/BaekJoon/Gold/G2_20183.java
package BaekJoon.Gold;

import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;

class Node implements Comparable<Node> {
    int node;
    long dist;

    public Node(int node, long dist) {
        this.node = node;
        this.dist = dist;
    }

    @Override
    public int compareTo(Node right) {
        if (dist != right.dist) {
            return Long.compare(dist, right.dist);
        }

        return 0;
    }
}

public class G2_20183 {
    public static boolean bfs(int n, long value, List<List<Node>> graph, int a, int b, long c) {
        // 거리 저장 배열 초기화
        List<Long> dist = new ArrayList<>();
        for (int i = 0; i < n + 1; i++) {
            dist.add(Long.MAX_VALUE);
        }

        Queue<Node> pque = new PriorityQueue<>();
        pque.add(new Node(a, 0L));

        dist.set(a, 0L);

        while (!pque.isEmpty()) {
            Node now = pque.poll();

            if (now.dist > dist.get(now.node))
                continue;

            for (int i = 0; i < graph.get(now.node).size(); i++) {
                Node next = graph.get(now.node).get(i);

                if (next.dist > value)
                    continue;

                long next_dist = now.dist + next.dist;

                if (next_dist > c)
                    continue;

                if (next_dist >= dist.get(next.node))
                    continue;

                pque.add(new Node(next.node, next_dist));
                dist.set(next.node, next_dist);
            }
        }

        if (dist.get(b) != Long.MAX_VALUE) {
            return true;
        } else {
            return false;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();
        int a = sc.nextInt();
        int b = sc.nextInt();
        long c = sc.nextLong();

        // 그래프 초기화
        List<List<Node>> graph = new ArrayList<>();
        ;
        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }

        // 그래프 입력
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            long cost = sc.nextLong();

            graph.get(u).add(new Node(v, cost));
            graph.get(v).add(new Node(u, cost));
        }

        long lo = 1;
        long hi = 1000000000;
        long answer = -1;

        while (lo <= hi) {
            long mid = (lo + hi) / 2;

            boolean res = bfs(n, mid, graph, a, b, c);

            if (res) {
                hi = mid - 1;
                answer = mid;
            } else {
                lo = mid + 1;
            }
        }

        System.out.println(answer);

        sc.close();
    }
}
